### DISCUSSION TITLE
Terraform support

### DISCUSSION THREAD

**Discussion by NotJustPizza** [timvisee/lazymc#5] [Ideas]
Add option to create/destroy external server using Terraform. https://github.com/PhilippeChepy/terraform-rs looks useful.

**Comment by NotJustPizza** [comment]
![lazymc drawio](https://user-images.githubusercontent.com/9921307/141650492-bef78763-fb14-40f7-92a9-5798260b58bb.png)

**Comment by timvisee** [comment]
Cool idea!

I assume you're suggesting to implement a Terraform script (not sure if it's called that) to spin up a fresh server in the cloud on connect. This is definitely something to look into at a later stage.

Luckily `lazymc` already supports querying, managing and proxying-to remote instances.

In the mean time, the probably already are scripts available to spin up a Minecraft server. You may be able to set this up by invoking the script as 'command' in `lazymc.toml`.

**Reply by NotJustPizza** [reply]
Terraform typically has many different files, so it's better to call it project instead of script, for example:
![image](https://user-images.githubusercontent.com/9921307/141681552-65ad5d1e-aa9d-4855-8129-71c4eaf4a29c.png)

I noticed there is support for custom `start-server`, but I didn't find anything allowing me to use custom `stop-server` script. I probably could catch term signal in `start-server` to trigger stop script, but it will be ugly solution without visibility.

Terraform supports variety of different APIs/clouds and allows us to run planning (previewing changes), creating and destroying. It shows number of changes planned/applied and their detailed descriptions. This data could be used by lazymc to, for example, report warnings if there are anomalies between runs. This kind of feature is something that would be more usable when directly supported instead of having workaround in bash.

In later stages it would be also nice to give some example configurations of Terraform for (a few) major clouds or even add them as "one-click" solutions bundled with lazymc for less technical people.

**Reply by timvisee** [reply]
> I noticed there is support for custom start-server, but I didn't find anything allowing me to use custom stop-server script. I probably could catch term signal in start-server to trigger stop script, but it will be ugly solution without visibility.

I agree.

Maybe the way server polling (and management) is done should be configurable.

With the current implementation, `lazymc` depends on two things:
- Server must be reachable over IP (status, ping)
- Start command must keep running until server dies

> (...) This kind of feature is something that would be more usable when directly supported instead of having workaround in bash.

Very true. Though I don't want to clutter the project with all kinds of extra things, this is probably one of those things worth implementing.

**Comment by NotJustPizza** [comment]
I'm going to prepare example Terraform setup for Vultr provider, but I need to wait for them to fix this issue first: https://github.com/vultr/terraform-provider-vultr/issues/176

**Reply by timvisee** [reply]
> I'm going to prepare example Terraform setup for Vultr provider, (..)

That would be fantastic!
